client/client.go

```diff
@@ -17,7 +34,41 @@ import (
 )
 
 type OpkClient struct {
-	Op OpenIdProvider
+	Op   OpenIdProvider
+	CosP *CosignerProvider
+}
+
+// Auth will attempt to authenticate to both the OpenID Provider and then the Cosigner.
+// If no Cosigner supplied it will default to OidcAuth
+func (o *OpkClient) Auth(
```

Comment:
Thats a really neat idea. If we can get it to work I think it would be a great improvement.

How does the redirect `CosignPkToken` trigger the redirect inside of `GetPKToken`? How does `GetPKToken` know to keep the HTTP session open so that the redir can be triggered?

Reply:
What I'm proposing also assumes that that code also has control over the server (I've made some comments about this elsewhere), so that people can plug in their own.

The redirect would happen as soon as the first call is done, and the triggering wouldn't necessarily require the HTTP session being left open, you can accomplish it with web page code too and a signal from the local server.

Reply:
@johncmerfeld  

Regarding `client.GetPKToken(...)` and `client.CosignPkToken( ...)` I'm not sure how this would work. I'm not saying is can't work but I don't see how to do it without opening a second browser window. Another issue is that depending on how the client is configured it may or may not need to do the redirect. I don't see how `client.GetPKToken(...) ` would know if it should do a redirect or not unless it was previously configured to expect the user to call `client.CosignPkToken( ...)`, but if the configuation makes calling `client.CosignPkToken( ...)` a requirement, why not just have one function which makes the appropriate calls based on the configuration.

> I don't think it's wise to have both Auth and OidcAuth in our API if we can help it. It violates the principle of having precisely one way to perform an action, which I would expect a library at v1.0 to adhere to.

Good call, that is exactly the plan!

`Auth` is the one call that does both or neither depending on configuration. `OidcAuth` exists for compatibility with code outside the project. Once this PR is merged, I'm going to make PRs into Docker's and SSHpiper's code to point them at `Auth` and then remove `OidcAuth`.

Reply:
Fair enough. I definitely don't have a good technical solution for you, and I agree that, given the constraints you've outlined, your approach makes sense!

Reply:
I would love a better solution, but I've been thinking about ways to do this since June 2023 and this is good enough.

Reply:
Agreed!

Reply:
I know this comment will never be read, but to put in my two cents here, I don't think committing to the other projects that use us is necessary since everything is version locked